### PR TITLE
[RISCV] Add back SiFive's cdiscard.d.l1, cflush.d.l1, and cease instructions.

### PR DESCRIPTION
**Pull request by topperc**
These were in LLVM 17 but removed from LLVM 18 due to an incorrect extension name being used.

This restores them with new extension names that match SiFive's downstream compiler. The extension name has been used internally for some time. It uses XSiFive instead of XSf like the newer extensions. `cease` did not have an internal extension name so its using the `XSf` convention.

The spec for the instructions is here https://sifive.cdn.prismic.io/sifive/767804da-53b2-4893-97d5-b7c030ae0a94_s76mc_core_complex_manual_21G3.pdf though the extension name is not listed.

Column width in the extension printing had to be changed to accommodate a longer extension name.

### CODE REVIEW

File: clang/test/Preprocessor/riscv-target-features.c [hunk 1]
```diff
@@ -60,6 +60,8 @@
 // CHECK-NOT: __riscv_xsfvfwmaccqqq {{.*$}}
 // CHECK-NOT: __riscv_xsfqmaccdod {{.*$}}
 // CHECK-NOT: __riscv_xsfvqmaccqoq {{.*$}}
+// CHECK-NOT: __riscv_sifivecdiscarddlone {{.*$}}
+// CHECK-NOT: __riscv_sifivecflushdlone {{.*$}}
```
Comment (dtcxzyw) [inline clang/test/Preprocessor/riscv-target-features.c:64]:
```suggestion
// CHECK-NOT: __riscv_xsifivecdiscarddlone {{.*$}}
// CHECK-NOT: __riscv_xsifivecflushdlone {{.*$}}
```

**Review by jrtc27 — DISMISSED** [review]
These need the vendor "sf." prefix

**Comment by jrtc27** [comment]
> By the way, is there any plan to support `CFLUSH.I.L1` in the future?

Flushing the instruction cache doesn't make sense given it can never be dirty. Invalidating/discarding does, but that's just what fence.i is doing?

**Comment by topperc** [comment]
@jrtc27 does this look better now?

**Comment by jrtc27** [comment]
> @jrtc27 does this look better now?

Yes; thanks

**Comment by apazos** [comment]
@topperc , how about the CEASE instruction? Can you add it in this patch?

**Comment by apazos** [comment]
Also, @topperc, we should add the new extensions  to the list of vendor extensions in the specs:
ttps://github.com/riscv-non-isa/riscv-toolchain-conventions/tree/master?tab=readme-ov-file#list-of-vendor-extensions

**Comment by quic-garvgupt** [comment]
Hi @topperc, can you add instruction alias for cflush and cdiscard instructions when the rs1 is X0 to `sf.cflush.d.l1`  and `sf.cflush.d.l1` respectively, as this register is optional according to spec?

**Comment by quic-garvgupt** [comment]
Also, I think we might need to update the extensions in the `RISCVProcessors.td` file under SIFIVE_S76 microcontroller?

**Comment by topperc** [comment]
> Hi @topperc, can you add instruction alias for cflush and cdiscard instructions when the rs1 is X0 to `sf.cflush.d.l1` and `sf.cflush.d.l1` respectively, as this register is optional according to spec?

x0 has special meaning, but the spec never says it is "optional".

**Comment by quic-garvgupt** [comment]
> > Also, I think we might need to update the extensions in the `RISCVProcessors.td` file under SIFIVE_S76 microcontroller?
> 
> This is a M-mode only extension, and we haven't historically been adding M or S mode extensions to the -mcpu lists. Except for `xiangshan-nanhu` having `Svinval`. But maybe there's a microcontroller vs application core distinction we should be making?

I see, I was not aware of this. Thanks for your reply

**Comment by topperc** [comment]
> > > Hi @topperc, can you add instruction alias for cflush and cdiscard instructions when the rs1 is X0 to `sf.cflush.d.l1` and `sf.cflush.d.l1` respectively, as this register is optional according to spec?
> > 
> > 
> > x0 has special meaning, but the spec never says it is "optional".
> 
> Possibly I am mistaken here but this is what I inferred from the below text in the mannual - rs1 is optional, so if a user does not specify it then by default it will be X0. Apologies if I am missing something here but just wanted to be clear.
> 
> From the mannual - `Opcode 0xFC000073, with **optional** rs1 field in bits [19:15]`

I think i had a bad copy of the file. There are several iterations. And one of the copies I had doesn't have the "optional" text on cflush.d.l1.

**Comment by quic-garvgupt** [comment]
Thanks for clearing the confusion around whether rs1 would be optional or not. Can we also add lit tests for the aliases?

**Comment by topperc** [comment]
> Thanks for clearing the confusion around whether rs1 would be optional or not. Can we also add lit tests for the aliases?

I already addded tests https://github.com/llvm/llvm-project/pull/83896/commits/f6f43e9f8ffa8b58d63178d28c826d0009de2f3b

**Comment by quic-garvgupt** [comment]
Thanks for the prompt reply and latest patchset.

1. Do we need to add documentation in RISCVUsage.rst file for xsfcease?
2. Also, as we are adding cease instruction in this PR, can we rename the PR to include the cease instruction as well?

**Comment by quic-garvgupt** [comment]
I believe the patch is ready to be merged now, but since I am not the reviewer here, someone else must approve it as well.

Thanks again for the patch!

**Review by dtcxzyw — APPROVED** [review]

**Pull request merged**